corehq/apps/app_manager/tests/test_suite_instances.py

```diff
@@ -260,17 +260,27 @@ def test_module_filter_instances_on_all_forms(self):
         self.module, self.form = factory.new_basic_module('m0', 'case1')
         self.module.module_filter = "instance('locations')/locations/"
         factory.new_form(self.module)
-        # two forms, two entries, two instances
-        self.assertXmlPartialEqual(
-            """
-            <partial>
-                <instance id='locations' src='jr://fixture/locations' />
-                <instance id='locations' src='jr://fixture/locations' />
-            </partial>
-            """,
-            factory.app.create_suite(),
-            "entry/instance"
-        )
+
+        suite = factory.app.create_suite()
+        instance_xml = "<partial><instance id='locations' src='jr://fixture/locations' /></partial>"
+        self.assertXmlPartialEqual(instance_xml, suite, "entry/command[@id='m0-f0']/../instance")
+        self.assertXmlPartialEqual(instance_xml, suite, "entry/command[@id='m0-f1']/../instance")
+
+    def test_module_filter_instances_on_all_forms_merged_modules(self):
+        # instances from module filters result in instance declarations in all forms in the module
+        # and if two modules are display_in_root, then on forms in the other module too
+        factory = AppFactory(build_version='2.20.0')  # enable_module_filtering
+        self.m0, self.m0f0 = factory.new_basic_module('m0', 'case1')
+        self.m0.put_in_root = True
+
+        self.m1, self.m1f0 = factory.new_basic_module('m1', 'case1')
+        self.m1.module_filter = "instance('locations')/locations/"
+        self.m1.put_in_root = True
+
+        suite = factory.app.create_suite()
+        instance_xml = "<partial><instance id='locations' src='jr://fixture/locations' /></partial>"
+        self.assertXmlPartialEqual(instance_xml, suite, "entry/command[@id='m0-f0']/..instance")
```

Comment:
I don't know the `..` syntax, what's that? I'm also uncertain what the `..` in `entry/command[@id='m0-f1']/../instance` is for - instances are direct children of commands, so is it needed?

Reply:
The `..` refers to the parent level in xpath, similar to `..` in unix filesystems.  I pulled the syntax from some other test files (I see it in `test_suite_case_tiles` and `test_shadow_forms`).  `instance` is actually a sibling of `command`, here's what this `entry` block looks like:

```xml
<entry>
  <command id="m1-f0">
    <text>
      <locale id="forms.m1f0"/>
    </text>
  </command>
  <instance id="locations" src="jr://fixture/locations"/>
</entry>
```
It's unfortunate that the `m1-f0` label is on the `command`, not the `entry`, as I don't think there's another good way to identify the entry you want (index, I guess, but that's not great).

I do notice I did `..instance` here and `../instance` above - both seem to work, but the latter seems more correct, so I'll update that.

Reply:
oh, I see, I must have misread the suite. Thanks for laying out an example.